### PR TITLE
provider/vsphere: govmomi v0.6.1 update and fixes

### PR DESCRIPTION
**Pull request by thetuxkeeper**
I used PR #6472 (thanks @catsby :) ), rebased it to master and started to fix some issues:
- datatypes
- `RemoveDevice`: "keepFiles" boolean was added => false for network devices (should be relevant for disks only)
- `CreateDisk`: not additional parameter for the datastore. `newDiskPath` logic removed ("empty" check is in govmomi now)

cc @chrislovecnm

### CODE REVIEW

**Comment by chrislovecnm** [comment]
LOL - we did the same stuff ... <facepalm>  ... I will review.

**Comment by chrislovecnm** [comment]
This is marked as WIP??

File: builtin/providers/vsphere/resource_vsphere_virtual_machine.go [hunk 1]
```diff
@@ -1692,7 +1684,7 @@ func (vm *virtualMachine) deployVirtualMachine(c *govmomi.Client) error {
 	for _, dvc := range devices {
 		// Issue 3559/3560: Delete all ethernet devices to add the correct ones later
 		if devices.Type(dvc) == "ethernet" {
-			err := newVM.RemoveDevice(context.TODO(), dvc)
+			err := newVM.RemoveDevice(context.TODO(), false, dvc)
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_machine.go]: We should consider setting `false` as a parameter.  No change needed at this point.

**Comment by chrislovecnm** [comment]
@dkalleg @markpeek - please review

@catsby @jen20 @stack72 - can one of you review if you have time?

**Comment by thetuxkeeper** [comment]
@chrislovecnm WIP was because I pushed it without much testing (had to leave the office)  and I wasn't completely sure about the function changes.

**Comment by jen20** [comment]
Ideally here someone with a vsphere setup would run the acceptance tests against this branch and ensure that there are no unexpected functional changes here - I just merged a fix from @thetuxkeeper into master, so it's possible it will want further rebasing. Thanks all!

**Comment by chrislovecnm** [comment]
@jen20 yes I will, just need someone with better golang kungfu to look at the casting that we are doing.

File: builtin/providers/vsphere/resource_vsphere_virtual_machine.go [hunk 2]
```diff
@@ -423,7 +423,7 @@ func resourceVSphereVirtualMachineUpdate(d *schema.ResourceData, meta interface{
 	configSpec := types.VirtualMachineConfigSpec{}
 
 	if d.HasChange("vcpu") {
-		configSpec.NumCPUs = d.Get("vcpu").(int)
+		configSpec.NumCPUs = int32(d.Get("vcpu").(int))
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_machine.go]: @jen20 the cast to `int32` is new ... opinion?  I know you go fu is strong :)

**Comment by thetuxkeeper** [comment]
I'll let the acceptance tests run against this branch with #6480 merged locally (it will fail without it).
@chrislovecnm or do you want to do it? Though it won't hurt if we both check the tests :)

**Comment by chrislovecnm** [comment]
Both of use need to.... lol

**Comment by chrislovecnm** [comment]
We have something odd going on:

```console
--- FAIL: TestAccVSphereVirtualMachine_basic (163.41s)
	testing.go:172: Step 0 error: Check failed: Check 5/9 error: vsphere_virtual_machine.foo: Attribute 'memory_reservation' expected "4096", got "0"
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/vsphere	163.431s
```
I may not have time to debug today ... But this needs to get resolved.

**Comment by chrislovecnm** [comment]
@thetuxkeeper do I have your email address?

**Comment by thetuxkeeper** [comment]
@chrislovecnm the error is fixed with #6480

**Comment by chrislovecnm** [comment]
<facepalm> ... Yah @markpeek just helped me out with that one

**Comment by chrislovecnm** [comment]
Looking better ... as @markpeek said we need to document how to run the darn integration tests better.  Maybe example bash script or something ...

**Comment by chrislovecnm** [comment]
@thetuxkeeper the only test that is failing is `TestAccVSphereVirtualMachine_ipv4Andipv6`, but that may be environment issues on my end.  Is it passing for you?  If so LGTM.

**Comment by thetuxkeeper** [comment]
OS customization doesn't work for me, so even `TestAccVSphereVirtualMachine_basic` fails. It seems my template is too minimalistic (minimal CentOS 7 installation), have to install some dependencies.

**Comment by chrislovecnm** [comment]
@thetuxkeeper you have open-vm-tools installed?

**Comment by chrislovecnm** [comment]
You will want the latest

**Comment by thetuxkeeper** [comment]
`TestAccVSphereVirtualMachine_basic` passed. `perl` was missing in the template. 
I just started the acceptance tests and had to restart. Forgot to remove the memory reservation => tests without setting it failed expecting it to be "0"

**Comment by chrislovecnm** [comment]
@jen20 @catsby @phinze @stack72 - might we get a merge? This is blocking a few different issues.  Integration tests are passing nicely.

**Comment by thetuxkeeper** [comment]
Most work, but some fail. Will look at it at friday (tomorrow we have a holiday and it's already late here)
```
=== RUN   TestAccVSphereVirtualMachine_diskInitType
--- FAIL: TestAccVSphereVirtualMachine_diskInitType (79.94s)
        testing.go:172: Step 0 error: Check failed: Check 9/10 error: vsphere_virtual_machine.thin: Attribute 'network_interface.#' expected "1", got "0"

=== RUN   TestAccVSphereVirtualMachine_ipv4Andipv6
--- FAIL: TestAccVSphereVirtualMachine_ipv4Andipv6 (96.15s)
        testing.go:172: Step 0 error: Check failed: Check 10/12 error: vsphere_virtual_machine.ipv4ipv6: Attribute 'network_interface.0.ipv4_gateway' expected "10.30.8.1", got ""
FAIL
```
Have to check `TestAccVSphereVirtualMachine_createWithExistingVmdk` too. It seems it deleted the vmdk and the next run it failed not finding the vmdk ... Do I have to copy the vmdk each time?
It also seems to fail to customize since it falls back to DHCP.
```
--- FAIL: TestAccVSphereVirtualMachine_createWithExistingVmdk (60.54s)
        testing.go:172: Step 0 error: After applying this step, the plan was not empty:

                DIFF:

                UPDATE: vsphere_virtual_machine.with_existing_vmdk
                  network_interface.0.ipv4_address:       "10.30.8.59" => "10.30.8.250"
                  network_interface.0.ipv4_prefix_length: "23" => "24"
```

**Comment by chrislovecnm** [comment]
@thetuxkeeper were are we on this?  @stack72 has mentioned that 0.7 will not arrive for a bit, and he is recommending that we push it.

Thoughts?

**Comment by thetuxkeeper** [comment]
@chrislovecnm : With #6656 my test run looks good.
If you can confirm that the tests are ok except the network interface fails sometimes, I think we can merge this one.
```
TF_ACC=1 go test ./builtin/providers/vsphere -v  -timeout 120m
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccVSphereFile_basic
Expected error received: cannot stat '[DS1] tf_file_test.vmdk': No such file
--- PASS: TestAccVSphereFile_basic (4.47s)
=== RUN   TestAccVSphereFile_renamePostCreation
Expected error received: cannot stat '[DS1] tf_test_file_moved.vmdk': No such file
Expected error received: cannot stat '[DS1] tf_test_file.vmdk': No such file
--- PASS: TestAccVSphereFile_renamePostCreation (7.60s)
=== RUN   TestAccVSphereFolder_basic
--- PASS: TestAccVSphereFolder_basic (3.31s)
=== RUN   TestAccVSphereFolder_nested
--- PASS: TestAccVSphereFolder_nested (3.20s)
=== RUN   TestAccVSphereFolder_dontDeleteExisting
--- PASS: TestAccVSphereFolder_dontDeleteExisting (3.71s)
=== RUN   TestAccVSphereVirtualMachine_basic
--- PASS: TestAccVSphereVirtualMachine_basic (97.74s)
=== RUN   TestAccVSphereVirtualMachine_diskInitType
--- PASS: TestAccVSphereVirtualMachine_diskInitType (108.73s)
=== RUN   TestAccVSphereVirtualMachine_dhcp
--- PASS: TestAccVSphereVirtualMachine_dhcp (93.50s)
=== RUN   TestAccVSphereVirtualMachine_custom_configs
--- PASS: TestAccVSphereVirtualMachine_custom_configs (85.98s)
=== RUN   TestAccVSphereVirtualMachine_createInExistingFolder
--- PASS: TestAccVSphereVirtualMachine_createInExistingFolder (85.49s)
=== RUN   TestAccVSphereVirtualMachine_createWithFolder
--- PASS: TestAccVSphereVirtualMachine_createWithFolder (94.29s)
=== RUN   TestAccVSphereVirtualMachine_createWithCdrom
--- PASS: TestAccVSphereVirtualMachine_createWithCdrom (97.25s)
=== RUN   TestAccVSphereVirtualMachine_createWithExistingVmdk
--- PASS: TestAccVSphereVirtualMachine_createWithExistingVmdk (69.68s)
=== RUN   TestAccVSphereVirtualMachine_updateMemory
--- PASS: TestAccVSphereVirtualMachine_updateMemory (139.98s)
=== RUN   TestAccVSphereVirtualMachine_updateVcpu
--- PASS: TestAccVSphereVirtualMachine_updateVcpu (153.46s)
=== RUN   TestAccVSphereVirtualMachine_ipv4Andipv6
--- FAIL: TestAccVSphereVirtualMachine_ipv4Andipv6 (76.18s)
        testing.go:234: Step 0 error: Check failed: Check 7/12 error: vsphere_virtual_machine.ipv4ipv6: Attribute 'network_interface.#' expected "1", got "0"
FAIL
exit status 1
FAIL    github.com/hashicorp/terraform/builtin/providers/vsphere        1124.593s
```

**Comment by chrislovecnm** [comment]
@stack72 @jen20 @phinze may we have a merge good sirs!!

**Comment by stack72** [comment]
Merging as requested :) Thanks all!

**Comment by chrislovecnm** [comment]
@stack72 thanks Paul

**Comment by thetuxkeeper** [comment]
@stack72 Thanks! :)

**Pull request merged**